### PR TITLE
[BALDE-17] Add Vector4 (x, y, z, w)

### PR DESCRIPTION
**Pull request by Blinkuu**
Added vec4 and fixed vec2, vec3.

### CODE REVIEW

File: src/math/vec4.h [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    }; // END OF CLASS Vec4
+
+        constexpr Vec4 Vec4::MakeUnitVec(const Vec4& vec) noexcept {
+                float len = vec.Length();
+                return Vec4(vec.GetX() / len, vec.GetY() / len, vec.GetZ() / len, vec.GetW() / len);
+        }
+
+        constexpr void Vec4::Normalize() noexcept {
+                float len = Length();
+                if (!len) {
```
Comment (4jeR) [inline src/math/vec4.h]: if len == 0 {
dividing
}

File: src/math/vec3.h [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+#pragma once
+
+#include <iostream>
+
+namespace Bald::Math {
+    class Vec3 {
+    public:
+        /**
+         * @fn                          Vec3
```
Comment (4jeR) [inline src/math/vec3.h]: * @fn                          Vec4

File: src/math/vec4.h [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+#pragma once
+
+#include <iostream>
+
+namespace Bald::Math {
+    class Vec4 {
+    public:
+        /**
+     * @fn                          Vec3
```
Comment (4jeR) [inline src/math/vec4.h]: * @fn                          Vec4

File: src/math/vec4.h [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        * @param [const Vec4&]         vec -> vector which will be used to create unit vector
+        * @return [Vec4]               unit long vector
+        */
+        [[nodiscard]] static constexpr Vec4 MakeUnitVec(const Vec4& vec) noexcept;
+
+        /**
+         * @fn                          Normalize
+         * @brief                       normalizes vector to the length of 1.0f
+         *                              (divides each component by the length of the vector)
+         * @return [Vec4&]              unit long vector
```
Comment (4jeR) [inline src/math/vec4.h]: no return value

File: src/math/vec4.h [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    }; // END OF CLASS Vec4
+
+        constexpr Vec4 Vec4::MakeUnitVec(const Vec4& vec) noexcept {
+                float len = vec.Length();
+                return Vec4(vec.GetX() / len, vec.GetY() / len, vec.GetZ() / len, vec.GetW() / len);
+        }
+
+        constexpr void Vec4::Normalize() noexcept {
+                float len = Length();
+                if (len) {
```
Comment (BIGbadEL) [inline src/math/vec4.h]: we agreed on using if(len != 0) convention.

File: src/math/vec4.h (hunk 1 above)
Comment (4jeR) [inline src/math/vec4.h]: if(len)

**Comment by piotrmoszkowicz** [comment]
Tests works without xvfb ?

**Comment by Blinkuu** [comment]
> Tests works without xvfb ?

Oui

**Comment by piotrmoszkowicz** [comment]
Hmmh, Travis must have changed something than. Good :) When I will have some spare time I will make the review.

File: src/math/vec2.h [hunk 6]
```diff
@@ -11,77 +11,77 @@ namespace Bald::Math {
...
-         * @param [float]               x variable
-         * @param [float]               y variable
-         */
-
+        * @fn                          Vec2
+        * @brief                       constructor
+        * @param [float]               x variable
+        * @param [float]               y variable
+        */
         constexpr Vec2(float x = 0.0f, float y = 0.0f)
```
Comment (BIGbadEL) [inline src/math/vec2.h]: explicit

File: src/math/vec2.h [hunk 7]
```diff
@@ -90,105 +90,164 @@ namespace Bald::Math {
...
 
         /**
          * @fn                          operator+=
          * @brief                       adds another vector to the current vector
          * @param [const Vec2&]         other vector
          * @return [Vec2&]              current vector added with the passed vector
          */
-        Vec2 &operator+=(const Vec2 &other) noexcept;
-
+        constexpr Vec2 &operator+=(const Vec2& other) noexcept;
```
Comment (BIGbadEL) [inline src/math/vec2.h]: formatting Vec2&

File: src/math/vec2.h [hunk 8]
```diff
@@ -90,105 +90,164 @@ namespace Bald::Math {
...
         /**
          * @fn                          operator*=
          * @brief                       multiplies current vector by the passed multiplier
-         * @param [const Vec2&]         other vector
+         * @param [const Vec23&]         other vector
          * @return [Vec2&]              current vector multiplied by the passed multiplier
          */
-        Vec2 &operator*=(float multiplier) noexcept;
-
+        constexpr Vec2& operator*=(float multiplier) noexcept;
```
Comment (BIGbadEL) [inline src/math/vec2.h:126]: there should be also *=(Vec2)

File: src/math/vec2.h [hunk 9]
```diff
@@ -90,105 +90,164 @@ namespace Bald::Math {
...
-    // END OF CLASS Vec2
-
-
-
-
+        float m_X;
+        float m_Y;
+    }; // END OF CLASS Vec2
+
+        constexpr Vec2 Vec2::MakeUnitVec(const Vec2& vec) noexcept {
```
Comment (BIGbadEL) [inline src/math/vec2.h]: what if length = 0 ?

File: src/math/vec2.h [hunk 10]
```diff
@@ -90,105 +90,164 @@ namespace Bald::Math {
...
+                float len = vec.Length();
+                return Vec2(vec.GetX() / len, vec.GetY() / len);
+        }
+
+        constexpr void Vec2::Normalize() noexcept {
+                float len = Length();
+                if (len) {
```
Comment (BIGbadEL) [inline src/math/vec2.h]: we agreed on format len != 0

**Review by BIGbadEL — COMMENTED** [review]
:)

File: src/math/vec2.h (hunk 10 above)
Comment (Blinkuu) [reply src/math/vec2.h]: Changed to match the convention.

File: src/math/vec2.h (hunk 9 above)
Comment (Blinkuu) [reply src/math/vec2.h]: Now returning VecX(0.0f,0.0f,0.0f)

File: src/math/vec2.h (hunk 8 above)
Comment (Blinkuu) [reply src/math/vec2.h:126]: lmao

File: src/math/vec2.h (hunk 7 above)
Comment (Blinkuu) [reply src/math/vec2.h]: Changed to match the convention.

File: src/math/vec2.h (hunk 6 above)
Comment (Blinkuu) [reply src/math/vec2.h]: Done

File: src/math/vec4.h (hunk 5 above)
Comment (Blinkuu) [reply src/math/vec4.h]: Changed to match the convention.

**Review by BIGbadEL — APPROVED** [review]

**Review by 4jeR — APPROVED** [review]

**Pull request merged**